src/stream-tcp-reassemble.c

```diff
@@ -1767,14 +1767,22 @@ int StreamTcpReassembleHandleSegment(ThreadVars *tv, TcpReassemblyThreadCtx *ra_
         dir = UPDATE_DIR_BOTH;
     }
 
+    const uint32_t base_seq = stream->base_seq;
+    const uint32_t isn = stream->isn;
     /* handle ack received */
     if ((dir == UPDATE_DIR_OPPOSING || dir == UPDATE_DIR_BOTH) &&
         StreamTcpReassembleHandleSegmentUpdateACK(tv, ra_ctx, ssn, opposing_stream, p) != 0)
     {
         SCLogDebug("StreamTcpReassembleHandleSegmentUpdateACK error");
         SCReturnInt(-1);
     }
-
+    /* The procedure above may swap content of ssn->server and ssn->client structures */
+    /* We have to continue with initial content of the stream in such case */
+    if (base_seq != stream->base_seq && isn != stream->isn && base_seq == opposing_stream->base_seq && isn == opposing_stream->isn)
```

Comment:
Why do we need the condition for `base_seq` ?
Is the `isn` test not enough ?

Reply:
What if the `isn` is the same on both sides? Small chance I suppose, except if the server would do it deliberately.

Reply:
Can we then copy the pointer to `stream` then and check it afterwards ?